### ISSUE TITLE
Making ComposableArchitecture dynamic

### ISSUE THREAD

**Issue by Atimca** [pointfreeco/swift-composable-architecture#298]
**Describe the bug**
Right now all provided libraries, including `ComposableArchitecture` itself are static. Which is really painful to deal with. TicTacToe example uses a hack with wrapping SPM into a dynamic framework.

**To Reproduce**
Connect `TCA` framework to the framework **B** and to the app **A**, framework **B** also suppose to be connected to the app **A**. 

> This approach will lead to the known issue: Swift package target is linked as a static library by '' and 4 other targets. This will result in duplication of library code.

**Expected behavior**
There's supposed to be a dynamic option of `TCA`

**Environment**
 - Xcode 12.0.1
 - Swift 5.3
 - OS  iOS 14

**Proposed solution**
Afaik with the new version Xcode this problem is not an issue anymore. I tried this solution in [my own branch](https://github.com/Atimca/swift-composable-architecture/commit/45cd2555e6df83fc9425868dc61bc3efc49bc18f)

```swift
  products: [
     .library(
       name: "ComposableArchitecture",
       type: .dynamic,
       targets: ["ComposableArchitecture"]
     )
```
And seems that it works!
<img width="703" alt="Screenshot 2020-09-28 at 18 49 03" src="https://user-images.githubusercontent.com/13145815/94462086-4995d380-01bb-11eb-873a-5a1a6deca17d.png">

I've seen some previous conversations about it as well. Not sure, what was the reason to remove dynamic variation from the framework. If it's needed to keep a static version, maybe it's a way to follow `KingFisher` [example](https://github.com/onevcat/Kingfisher/blob/master/Package.swift). They provide two versions of the library static and dynamic one.

**Comment by stephencelis** [comment]
> TicTacToe example uses a hack with wrapping SPM into a dynamic framework.

According to some Apple engineers we spoke with, this is in fact the way to go right now for incorporating packages into a project. We've tried to document this in the README, although it is brief: https://github.com/pointfreeco/swift-composable-architecture#installation

Because we're dealing with an Apple bug here we think that introducing a `.dynamic` variant would be a hack of its own and complicate the library: we'd need to introduce an extra library name per target (like KingFisher does), document the workaround/bug, make all of TCA's upstream dependencies have dynamic variants, etc. The KingFisher PR links to some Swift forums discussion: https://github.com/onevcat/Kingfisher/pull/1426

One simplification we've experimented with in an application we're building is to power it with a single local Swift package that declares all of its dependencies in one place. Then the application can embed this single static library. It's worked well in Xcode 12, where some bugs around SPM, previews, and playgrounds were fixed.

**Comment by Atimca** [comment]
Thank you for the fast reply @stephencelis. Hope we'll find the proper solution soon. Also, not 100% related to TCA itself, but to the package of it and maybe it's good to share it in this issue. I bumped in the problem where I have my local SPMs where I wanted to connect Swift CasePath library which is a part of TCA itself. Because of the static nature and because TCA is the remote library I can't easily wrap casepathes. I can connect wrapper over TCA to it, but I'd like to keep my local small dependencies as clean as possible.

**Comment by stephencelis** [comment]
@Atimca As long as you do all the dependency work in a single place (like the local package trick I mentioned), you should be able to introduce a TCA and a separate dependency that depends on swift-case-paths without the duplication error.

**Comment by stephencelis** [comment]
@Atimca Since this is an SPM bug/limitation that's being tracked, can we close this issue out?

**Comment by Atimca** [comment]
Sure @stephencelis I've closed it. Thx again for the help ✌️

**Comment by andriyslyusar** [comment]
@stephencelis I have a similar problem, so decided not to open a new issue and continue the thread.

I am trying to implement next setup:
1. SPM feature package (A1) has a dependency to TCA.
2. Main app has dependency to local A1 package and TCA.

Any idea can I somehow make everything compile? In general I am trying to implement modular application, but instead of targets use SPM packages. Thanks in advance.

**Comment by stephencelis** [comment]
@andriyslyusar If you have a local A1 package that depends on TCA your main app should be able to depend on A1 and automatically pick up both packages. Wanna give that a shot?

**Comment by andriyslyusar** [comment]
Thanks a lot, it works exactly as you described.
